diff --git a/src/polysolve/nonlinear/line_search/Backtracking.cpp b/src/polysolve/nonlinear/line_search/Backtracking.cpp
index da8ee66..8b3e67e 100644
--- a/src/polysolve/nonlinear/line_search/Backtracking.cpp
+++ b/src/polysolve/nonlinear/line_search/Backtracking.cpp
@@ -77,7 +77,7 @@ namespace polysolve::nonlinear::line_search
         {
             TVector new_grad;
             objFunc.gradient(new_x, new_grad);
-            return new_grad.norm() < old_grad.norm(); // TODO: cache old_grad.norm()
+            return new_grad.norm() < old_grad.norm() || new_energy < old_energy; // TODO: cache old_grad.norm()
         }
         return new_energy < old_energy;
     }
diff --git a/src/polysolve/nonlinear/line_search/LineSearch.cpp b/src/polysolve/nonlinear/line_search/LineSearch.cpp
index a9be577..1f8bdc6 100644
--- a/src/polysolve/nonlinear/line_search/LineSearch.cpp
+++ b/src/polysolve/nonlinear/line_search/LineSearch.cpp
@@ -19,6 +19,51 @@
 
 namespace polysolve::nonlinear::line_search
 {
+    namespace
+    {
+        template <typename T>
+std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
+            for (const auto& i : v) os << i << " ";
+            return os;
+        }
+
+        void plot_over_descent_direction(
+            Problem& objFunc,
+            const Eigen::VectorXd& x,
+            const Eigen::VectorXd& dx,
+            double max_step_size = 1)
+        {
+            double alpha = 1e-20;
+            std::vector<double> alphas, values, linearApprox;
+            objFunc.solution_changed(x);
+            alphas.push_back(0);
+            values.push_back(objFunc.value(x));
+            linearApprox.push_back(objFunc.value(x));
+            Eigen::VectorXd grad;
+            objFunc.gradient(x, grad);
+            while (alpha < max_step_size)
+            {
+                Eigen::VectorXd y = x + dx * alpha;
+                objFunc.solution_changed(y);
+
+                alphas.push_back(alpha);
+                values.push_back(objFunc.value(y));
+                linearApprox.push_back(values[0] + grad.dot(dx) * alpha);
+
+                alpha *= 1.5;
+            }
+
+            objFunc.solution_changed(x);
+
+            std::cout << std::setprecision(20) << std::fixed << "x = txt_to_array(\"" << x.transpose() << "\")\n";
+            std::cout << std::setprecision(20) << std::fixed << "g = txt_to_array(\"" << grad.transpose() << "\")\n";
+            std::cout << std::setprecision(20) << std::fixed << "dx = txt_to_array(\"" << dx.transpose() << "\")\n";
+            std::cout << std::setprecision(20) << std::fixed << "a = txt_to_array(\"" << alphas << "\")\n";
+            std::cout << std::setprecision(20) << std::fixed << "y = txt_to_array(\"" << values << "\")\n";
+            std::cout << std::setprecision(20) << std::fixed << "approx = txt_to_array(\"" << linearApprox << "\")\n";
+        }
+    }
+
     static constexpr double NaN = std::numeric_limits<double>::quiet_NaN();
 
     std::shared_ptr<LineSearch> LineSearch::create(const json &params, spdlog::logger &logger)
@@ -145,6 +190,7 @@ namespace polysolve::nonlinear::line_search
             step_size = compute_descent_step_size(x, delta_x, objFunc, use_grad_norm, initial_energy, initial_grad, step_size);
             if (std::isnan(step_size))
             {
+                // plot_over_descent_direction(objFunc, x, delta_x, collision_free_step_size);
                 // Superclass::save_sampled_values("failed-line-search-values.csv", x, delta_x, objFunc);
                 return NaN;
             }
@@ -166,6 +212,7 @@ namespace polysolve::nonlinear::line_search
             assert(abs(initial_energy - objFunc(x)) < 1e-15);
 
             objFunc.line_search_end();
+            // plot_over_descent_direction(objFunc, x, delta_x, collision_free_step_size);
             return NaN;
         }
 
